src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -0,0 +1,354 @@
+        match tag {
+            TAG_MESSAGE => Ok(Tag::Message),
+            TAG_PUSH => Ok(Tag::Push),
+            TAG_REKEY => Ok(Tag::Rekey),
+            TAG_FINAL => Ok(Tag::Final),
+            _ => Err(())
+        }
+    }
+}
+
+new_type! {
+    /// `Key` for symmetric encryption.
+    ///
+    /// When a `Key` goes out of scope its contents will be overwritten in
+    /// memory.
+    secret Key(KEYBYTES);
+}
+
+new_type! {
+    /// An encrypted stream starts with a short header, whose size is HEADERBYTES bytes.
+    /// That header must be sent/stored before the sequence of encrypted messages,
+    /// as it is required to decrypt the stream.
+    public Header(HEADERBYTES);
+}
+
+impl Key {
+    /// Randomly generates a key for authenticated encryption.
+    ///
+    /// THREAD SAFETY: this method is safe provided that you have called
+    /// `sodiumoxide::init()` once before using any other function from
+    /// sodiumoxide.
+    // TODO: create a new `new_type!` macro for keys. It will probably look like
+    // `public`, and then just have this method.
+    pub fn new() -> Key {
+        let mut key: [u8; KEYBYTES] = unsafe { mem::uninitialized() };
+        randombytes_into(&mut key);
+        Key(key)
+    }
+}
+
+/// `Encryptor` contains the state for multi-part (streaming) computations. This
+/// allows the caller to process encryption of a sequence of multiple messages.
+pub struct Encryptor($state_name);
+
+impl Encryptor {
+    /// Initializes an `Encryptor` using a provided `key`. Returns the
+    /// `Encryptor` object and a `Header`, which is needed by the recipient to
+    /// initialize a corresponding `Decryptor`. The `key` will not be needed be
+    /// required for any subsequent authenticated encryption operations.
+    /// If you would like to securely generate a key and initialize an
+    /// `Encryptor` at the same time see the `new` method.
+    /// Network protocols can leverage the key exchange API in order to get a
+    /// shared key that can be used to encrypt streams. Similarly, file
+    /// encryption applications can use the password hashing API to get a key
+    /// that can be used with the functions below.
+    pub fn init(key: &Key) -> Result<(Self, Header), ()> {
+        let mut header: [u8; HEADERBYTES] = unsafe { mem::uninitialized() };
+        let mut state: $state_name = unsafe { mem::uninitialized() };
+
+        let rc = unsafe {
+            $init_push_name(&mut state, header.as_mut_ptr(), key.0.as_ptr())
+        };
+        if rc != 0 {
+            return Err(());
+        }
+
+        Ok((Encryptor(state), Header(header)))
+    }
+
+
+    /// Securely generates a key and uses it to initialize an `Encryptor`.
+    /// Returns the `Encryptor` object, a `Header` (which is needed by the
+    /// recipient to initialize a corresponding `Decryptor`), and the `Key`
+    /// object.
+    pub fn new() -> Result<(Self, Header, Key), ()> {
+        let key = Key::new();
```

Comment:
I think we might skip this methid completely. There is already a way to initialize a random key, and there is a way to create an Encryptor from a key, so this one feels superfluous.

Reply:
Keeping as a convenience method.